### PR TITLE
[#157148060] Login with PIN screen

### PR DESCRIPTION
**Pull request by ClaudioConversa**
Add a screen to login if PIN was set during onBoarding

screen shot 2018-06-04 at 12 27 32

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#157148060](https://www.pivotaltracker.com/story/show/157148060): Come CIT, quando apro l'APP e ho già concluso l'Onboarding  devo inserire il PIN prima di accedere allo screen principale.

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: ts/theme/variables.ts [hunk 1]
```diff
@@ -38,6 +38,7 @@ const customVariables = Object.assign(materialVariables, {
   brandDanger: "#CC3333",
   brandLightGray: "#E6E9F2",
   brandDarkGray: "#5C6F82",
+  brandWhite: "#FFFFFF",
```
Comment (gunzip) [inline ts/theme/variables.ts]: I guess you can just use `colorWhite` which is defined below

File: ts/components/Pinpad.tsx [hunk 2]
```diff
@@ -8,6 +8,9 @@ import { PIN_LENGTH } from "../utils/constants";
 type OwnProps = {
   autofocus: boolean;
   compareWithCode?: string;
+  inactiveColor?: string;
```
Comment (cloudify) [inline ts/components/Pinpad.tsx]: looks like `inactiveColor` and `activeColor` shouldn't be optional

File: ts/components/Pinpad.tsx [hunk 3]
```diff
@@ -29,14 +39,15 @@ const Pinpad: React.SFC<Props> = props => {
       codeLength={PIN_LENGTH}
       compareWithCode={compareWithCode || ""}
       cellBorderWidth={2}
-      inactiveColor={color(variables.brandLightGray)
+      inactiveColor={color(inactiveColor || variables.brandLightGray)
```
Comment (cloudify) [inline ts/components/Pinpad.tsx]: I would make `activeColor` and `inactiveColor` required so you don't have to resort to referring to the `variables`, let the parent component decide

File: ts/sagas/index.ts [hunk 4]
```diff
@@ -8,6 +8,7 @@ import { all, Effect, fork } from "redux-saga/effects";
 import mainSaga from "./main";
 import notificationsSaga from "./notifications";
 import onboardingSaga from "./onboarding";
+import pinloginSaga from "./pinlogin";
```
Comment (cloudify) [inline ts/sagas/index.ts]: `pinLoginSaga`

File: ts/sagas/pinlogin.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import { PinValidateRequest } from "../store/actions/pinlogin";
+import { getPin } from "../utils/keychain";
+
+/**
+ * The PIN step of the pin login
+ */
+function* pinLoginSaga(): Iterator<Effect> {
+  const navigateToPinLoginNavigatorAction = NavigationActions.navigate({
+    routeName: ROUTES.PINLOGIN,
+    key: undefined
```
Comment (cloudify) [inline ts/sagas/pinlogin.ts]: what does `key: undefined` do?

File: ts/sagas/pinlogin.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  });
+  yield put(navigateToPinLoginNavigatorAction);
+}
+function* validatePinSaga(action: PinValidateRequest): Iterator<Effect> {
```
Comment (cloudify) [inline ts/sagas/pinlogin.ts]: add an empty line between the functions

File: locales/en.json [hunk 7]
```diff
@@ -106,6 +106,14 @@
       "confirmInvalid": "the two pins do not match"
     }
   },
+  "pin_login":{
+    "pin": {
+      "pinInfo": "Put the fingerprint on the sensor or enter the PIN",
```
Comment (cloudify) [inline locales/en.json]: I would wait until we have the fingerprint functionality before adding it to the message - besides, the fingerprint login will not be available on all devices

File: locales/it.json [hunk 8]
```diff
@@ -105,6 +105,14 @@
       "confirmInvalid": "i due pin non coincidono"
     }
   },
+  "pin_login":{
+    "pin": {
+      "pinInfo": "Appoggia l'impronta sul sensore oppure inserisci il codice PIN",
```
Comment (cloudify) [inline locales/it.json]: same here, remove the fingerprint stuff until we implement it

File: ts/sagas/pinlogin.ts (hunk 5 above)
Comment (ClaudioConversa) [reply ts/sagas/pinlogin.ts]: key is an optional parameter string or null,  if set, navigation will go back from the given key. If null, navigation will go back anywhere.

File: ts/navigation/PinNavigator.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import { StackNavigator } from "react-navigation";
+
+import PinLoginScreen from "../screens/PinLoginScreen";
+import ROUTES from "./routes";
+
+/**
+ * The onboarding related stack of screens of the application.
+ */
+const navigator = StackNavigator(
```
[the remark this answers is not in the document]
Comment (cloudify) [reply ts/navigation/PinNavigator.ts]: (this matter has been discussed in the weekly sprint review)

File: ts/navigation/routes.ts [hunk 10]
```diff
@@ -31,6 +31,9 @@ const ROUTES = {
   WALLET_ADD_PAYMENT_METHOD: "WALLET_ADD_PAYMENT_METHOD",
   WALLET_ADD_CARD: "WALLET_ADD_CARD",
 
+  // Pin
+  PINLOGIN: "PINLOGIN",
```
Comment (cloudify) [inline ts/navigation/routes.ts]: `PIN_LOGIN` (for consistency with snake case of other route names)

File: ts/store/reducers/pinlogin.ts [hunk 11]
```diff
@@ -0,0 +1,31 @@
+/**
+ * A reducer for the PinLogin.
+ * @flow
+ */
+
+import { PIN_LOGIN_VALIDATE_FAILURE } from "../actions/constants";
+import { Action } from "../actions/types";
+
+export type PinLoginState = Readonly<{
+  PinConfirmed: string;
```
Comment (cloudify) [inline ts/store/reducers/pinlogin.ts]: instead of `string`, the type of `PinConfirmed` should be `"PinCheck" | "PinConfirmedInvalid"`

**Comment by cloudify** [comment]
@fpersico please review

File: ts/sagas/pinlogin.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
+/**
+ * A saga that manages the Onboarding.
```
Comment (fpersico) [inline ts/sagas/pinlogin.ts]: Onboarding?

File: ts/sagas/pinlogin.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+  } catch (error) {
+    yield put({
+      type: PIN_LOGIN_VALIDATE_FAILURE
+    });
+  }
+}
+
+export default function* root(): Iterator<Effect> {
+  yield takeLatest(PIN_LOGIN_INITIALIZE, pinLoginSaga);
+  yield takeLatest(PIN_LOGIN_VALIDATE_REQUEST, validatePinSaga);
```
Comment (fpersico) [inline ts/sagas/pinlogin.ts]: validatePinSaga -> pinValidateSaga

File: ts/sagas/session.ts [hunk 14]
```diff
@@ -56,7 +56,6 @@ function* landingStep(): Iterator<Effect> {
 function* sessionSaga(): Iterator<Effect> {
```
Comment (fpersico) [inline ts/sagas/session.ts]: This file is not related to the PR

File: ts/store/reducers/types.ts [hunk 15]
```diff
@@ -2,12 +2,14 @@ import { NavigationState } from "react-navigation";
 import { FormStateMap } from "redux-form";
 
 import { Action } from "../actions/types";
+
```
Comment (fpersico) [inline ts/store/reducers/types.ts]: Why a new line?
